File: src/ML2Typescript/Examples/MSetAVL.ml
type comparaison =
  | Eq
  | Lt
  | Gt

type 'a tree =
  | Empty
  | Node of int * 'a tree * 'a * 'a tree

let check (res: int): comparaison =
  if res = 0 then Eq
  else (if res < 0 then Lt else Gt);;

let max (i: int) (j: int): int =
  if i > j then i else j;;

let getOrElse (opt: 'a option) (other: 'a): 'a =
  match opt with
  | None -> other
  | Some v -> v;;

let is_empty (t: 'a tree) : bool = match t with
  | Empty -> true
  | Node (_, _, _, _) -> false;;

let rec mem (x: 'a) (t: 'a tree) (cmp: ('a -> 'a -> int)): bool =
  match t with
  | Empty -> false
  | Node (_, l, k, r) ->
    (match (check (cmp x k)) with
     | Eq -> true
     | Lt -> (mem x l cmp)
     | Gt -> (mem x r cmp));;

let rec find (x: 'a) (t: 'a tree) (cmp: ('a -> 'a -> int)): 'a option =
  match t with
  | Empty -> None
  | Node (_, l, k, r) ->
    (match (check (cmp x k)) with
     | Eq -> Some k
     | Lt -> (find x l cmp)
     | Gt -> (find x r cmp));;

let rec min_elt (t: 'a tree): 'a option = match t with
  | Empty -> None
  | Node (_, l, x, _) ->
    (match l with
     | Empty -> Some x
     | Node (_, _, _, _) -> (min_elt l));;

let rec max_elt (t: 'a tree): 'a option = match t with
  | Empty -> None
  | Node (_, _, x, r) ->
    (match r with
     | Empty -> Some x
     | Node (_, _, _, _) -> (max_elt r));;

let height (t: 'a tree): int = match t with
  | Empty -> 0
  | Node (h, _, _, _) -> h;;

let rec size (t: 'a tree): int = match t with
  | Empty -> 0
  | Node (h, l, _, r) -> (size l) + (size r) + 1;;

let create (l:'a tree) (x: 'a) (r: 'a tree): 'a tree =
  Node ((max (height l) (height r)) + 1, l, x, r);;

let leaf (x: 'a): 'a tree = (create Empty x Empty);;

let bal (l:'a tree) (x: 'a) (r: 'a tree): 'a tree =
  let hl: int = (height l) in
  let hr: int = (height r) in
  if (hr + 2) < hl
  then (match l with
      | Empty -> raise Exit
      | Node (_, ll, lx, lr) ->
        if (height lr) <= (height ll)
        then (create ll lx (create lr x r))
        else (match lr with
            | Empty -> raise Exit
            | Node (_, lrl, lrx, lrr) ->
              (create (create ll lx lrl) lrx (create lrr x r))))
  else if (hl + 2) < hr
  then (match r with
      | Empty -> raise Exit
      | Node (_, rl, rx, rr) ->
        if (height rl) <= (height rr)
        then (create (create l x rl) rx rr)
        else (match rl with
            | Empty -> raise Exit
            | Node (_, rll, rlx, rlr) ->
              (create (create l x rll) rlx (create rlr rx rr))))
  else (create l x r);;

let rec add (x:'a) (t: 'a tree) (cmp: ('a -> 'a -> int)): 'a tree =
  match t with
  | Empty -> (leaf x)
  | Node (h, l, y, r) ->
    (match (check (cmp x y)) with
     | Eq -> (create l y r)
     | Lt -> (bal (add x l cmp) y r)
     | Gt -> (bal l y (add x r cmp)));;

let rec remove_min (l: 'a tree) (x:'a) (r: 'a tree): ('a tree * 'a) =
  match l with
  | Empty -> (r,x)
  | Node (_, ll, lx, lr) ->
    let (l2, m) = (remove_min ll lx lr) in ((bal l2 x r),m);;

let merge (s1: 'a tree) (s2: 'a tree): 'a tree =
  match s1 with
  | Empty -> s2
  | Node (_, _, _, _) ->
    (match s2 with
     | Empty -> s1
     | Node (_, l2, x2, r2) ->
       let (s3, m) = (remove_min l2 x2 r2) in (bal s1 m s3));;

let rec remove (x: 'a) (t: 'a tree) (cmp: ('a -> 'a -> int)): 'a tree =
  match t with
  | Empty -> Empty
  | Node (_, l, y, r) ->
    (match (check (cmp x y)) with
     | Eq -> (merge l r)
     | Lt -> (bal (remove x l cmp) y r)
     | Gt -> (bal l y (remove x r cmp)));;

let rec toList (t: 'a tree): 'a list =
  match t with
  | Empty -> []
  | Node (_, l, y, r) -> (toList l) @ (y :: (toList r));;

File: src/ML2Typescript/Examples/TS/MSetAVL.ts
import { Option, none, some } from './Lib.js';
import * as ArrayList from './ArrayList.js';

type Comparaison =
    | { tag: 'eq' }
    | { tag: 'lt' }
    | { tag: 'gt' }

function eq(): Comparaison {
    return { tag: 'eq' };
}
function lt(): Comparaison {
    return { tag: 'lt' };
}
function gt(): Comparaison {
    return { tag: 'gt' };
}


type Tree<A> =
    | { tag: 'empty' }
    | { tag: 'node', content: [number, Tree<A>, A, Tree<A>] }

function empty<A>(): Tree<A> {
    return { tag: 'empty' };
}
function node<A>(arg1: number, arg2: Tree<A>, arg3: A, arg4: Tree<A>): Tree<A> {
    return { tag: 'node', content: [arg1, arg2, arg3, arg4] };
}


function check(res: number): Comparaison {
    if (res == 0) {
        return eq();
    } else {
        if (res < 0) {
            return lt();
        } else {
            return gt();
        }
    }
}

function max(i: number, j: number): number {
    if (i > j) {
        return i;
    } else {
        return j;
    }
}

function getOrElse<A>(opt: Option<A>, other: A): A {
    switch (opt.tag) {
        case 'none':
            return other;
        case 'some':
            const { content: [v] } = opt;
            return v;
    }
}

function is_empty<A>(t: Tree<A>): boolean {
    switch (t.tag) {
        case 'empty':
            return true;
        case 'node':
            const { content: [, , , ] } = t;
            return false;
    }
}

function mem<A>(x: A, t: Tree<A>, cmp: (_: A, __: A) => number): boolean {
    switch (t.tag) {
        case 'empty':
            return false;
        case 'node':
            const { content: [, l, k, r] } = t;
            switch (check(cmp(x, k)).tag) {
                case 'eq':
                    return true;
                case 'lt':
                    return mem(x, l, cmp);
                case 'gt':
                    return mem(x, r, cmp);
            }
    }
}

function find<A>(x: A, t: Tree<A>, cmp: (_: A, __: A) => number): Option<A> {
    switch (t.tag) {
        case 'empty':
            return none();
        case 'node':
            const { content: [, l, k, r] } = t;
            switch (check(cmp(x, k)).tag) {
                case 'eq':
                    return some(k);
                case 'lt':
                    return find(x, l, cmp);
                case 'gt':
                    return find(x, r, cmp);
            }
    }
}

function min_elt<A>(t: Tree<A>): Option<A> {
    switch (t.tag) {
        case 'empty':
            return none();
        case 'node':
            const { content: [, l, x, ] } = t;
            switch (l.tag) {
                case 'empty':
                    return some(x);
                case 'node':
                    const { content: [, , , ] } = l;
                    return min_elt(l);
            }
    }
}

function max_elt<A>(t: Tree<A>): Option<A> {
    switch (t.tag) {
        case 'empty':
            return none();
        case 'node':
            const { content: [, , x, r] } = t;
            switch (r.tag) {
                case 'empty':
                    return some(x);
                case 'node':
                    const { content: [, , , ] } = r;
                    return max_elt(r);
            }
    }
}

function height<A>(t: Tree<A>): number {
    switch (t.tag) {
        case 'empty':
            return 0;
        case 'node':
            const { content: [h, , , ] } = t;
            return h;
    }
}

function size<A>(t: Tree<A>): number {
    switch (t.tag) {
        case 'empty':
            return 0;
        case 'node':
            const { content: [h, l, , r] } = t;
            return ((size(l) + size(r)) + 1);
    }
}

function create<A>(l: Tree<A>, x: A, r: Tree<A>): Tree<A> {
    return node((max(height(l), height(r)) + 1), l, x, r);
}

function leaf<A>(x: A): Tree<A> {
    return create(empty(), x, empty());
}

function bal<A>(l: Tree<A>, x: A, r: Tree<A>): Tree<A> {
    const hl: number = height(l);
    const hr: number = height(r);
    if ((hr + 2) < hl) {
        switch (l.tag) {
            case 'empty':
                throw new Error()
            case 'node':
                const { content: [, ll, lx, lr] } = l;
                if (!(height(lr) > height(ll))) {
                    return create(ll, lx, create(lr, x, r));
                } else {
                    switch (lr.tag) {
                        case 'empty':
                            throw new Error()
                        case 'node':
                            const { content: [, lrl, lrx, lrr] } = lr;
                            return create(create(ll, lx, lrl), lrx, create(lrr, x, r));
                    }
                }
        }
    } else {
        if ((hl + 2) < hr) {
            switch (r.tag) {
                case 'empty':
                    throw new Error()
                case 'node':
                    const { content: [, rl, rx, rr] } = r;
                    if (!(height(rl) > height(rr))) {
                        return create(create(l, x, rl), rx, rr);
                    } else {
                        switch (rl.tag) {
                            case 'empty':
                                throw new Error()
                            case 'node':
                                const { content: [, rll, rlx, rlr] } = rl;
                                return create(create(l, x, rll), rlx, create(rlr, rx, rr));
                        }
                    }
            }
        } else {
            return create(l, x, r);
        }
    }
}

function add<A>(x: A, t: Tree<A>, cmp: (_: A, __: A) => number): Tree<A> {
    switch (t.tag) {
        case 'empty':
            return leaf(x);
        case 'node':
            const { content: [h, l, y, r] } = t;
            switch (check(cmp(x, y)).tag) {
                case 'eq':
                    return create(l, y, r);
                case 'lt':
                    return bal(add(x, l, cmp), y, r);
                case 'gt':
                    return bal(l, y, add(x, r, cmp));
            }
    }
}

function remove_min<A>(l: Tree<A>, x: A, r: Tree<A>): [Tree<A>, A] {
    switch (l.tag) {
        case 'empty':
            return [r, x];
        case 'node':
            const { content: [, ll, lx, lr] } = l;
            const [l2, m] = remove_min(ll, lx, lr);
            return [bal(l2, x, r), m];
    }
}

function merge<A>(s1: Tree<A>, s2: Tree<A>): Tree<A> {
    switch (s1.tag) {
        case 'empty':
            return s2;
        case 'node':
            const { content: [, , , ] } = s1;
            switch (s2.tag) {
                case 'empty':
                    return s1;
                case 'node':
                    const { content: [, l2, x2, r2] } = s2;
                    const [s3, m] = remove_min(l2, x2, r2);
                    return bal(s1, m, s3);
            }
    }
}

function remove<A>(x: A, t: Tree<A>, cmp: (_: A, __: A) => number): Tree<A> {
    switch (t.tag) {
        case 'empty':
            return empty();
        case 'node':
            const { content: [, l, y, r] } = t;
            switch (check(cmp(x, y)).tag) {
                case 'eq':
                    return merge(l, r);
                case 'lt':
                    return bal(remove(x, l, cmp), y, r);
                case 'gt':
                    return bal(l, y, remove(x, r, cmp));
            }
    }
}

function toList<A>(t: Tree<A>): A[] {
    switch (t.tag) {
        case 'empty':
            return [];
        case 'node':
            const { content: [, l, y, r] } = t;
            return [...toList(l), ...[y, ...toList(r)]];
    }
}
